tsfc/coffee_mode.py

```diff
@@ -0,0 +1,317 @@
+            all_atomics.append(_atomics)
+            all_rest.append(monomial.rest)
+        # Create new MonomialSum for the factorised out terms
+        sub_monomial_sum = MonomialSum()
+        for _atomics, _rest in zip(all_atomics, all_rest):
+            sub_monomial_sum.add((), _atomics, _rest)
+        sub_monomial_sum = optimise_monomial_sum(sub_monomial_sum, argument_indices)
+        assert len(sub_monomial_sum) > 0
+        if len(sub_monomial_sum) == 1:
+            # result is a product, add to new MonomialSum directly
+            sub_monomial, = sub_monomial_sum
+            new_atomics = sub_monomial.atomics
+            new_atomics += (oa,)  # add back common factor
+            new_rest = sub_monomial.rest
+        else:
+            # result is a sum, we need to create new node
+            new_node = monomial_sum_to_expression(sub_monomial_sum)
+            new_atomics = [oa]
+            new_rest = one
+            if set(argument_indices) & set(new_node.free_indices):
+                new_atomics.append(new_node)
+            else:
+                new_rest = new_node
+        new_monomial_sum.add(sum_indices, new_atomics, new_rest)
+    return new_monomial_sum
+
+
+def optimise_monomial_sum(monomial_sum, argument_indices):
+    """Choose optimal common atomic subexpressions and factorise a
+    :class:`MonomialSum` object.
+
+    :arg monomial_sum: a :class:`MonomialSum` object
+    :arg argument_indices: tuple of argument indices
+
+    :returns: factorised `MonomialSum` object
+    """
+    all_optimal_atomics = []  # [(sum_indces, optimal_atomics)]
+    for sum_indices in unique_sum_indices(monomial_sum):
+        # throw away other atomics here
+        optimal_atomics, _ = find_optimal_atomics(monomial_sum, frozenset(sum_indices), argument_indices)
+        all_optimal_atomics.extend([(sum_indices, atomic) for atomic in optimal_atomics])
+    # This algorithm is O(N!), where N = len(optimal_atomics)
+    # we could truncate the optimal_atomics list at say 10
+    return factorise_atomics(monomial_sum, all_optimal_atomics, argument_indices)
+
+
+def count_flop(expression):
+    """Count the total floating point operations required to compute a GEM node.
+    This function assumes that all subnodes that occur more than once induce a
+    temporary, and are therefore only computed once.
+
+    :arg expression: GEM expression
+
+    :returns: total number of FLOPs to compute the GEM expression
+    """
+    flop = 0
+    for node in traversal([expression]):
+        if isinstance(node, (Sum, Product, Division, MathFunction, Comparison, Power)):
+            flop += numpy.prod([idx.extent for idx in node.free_indices])
+        elif isinstance(node, IndexSum):
+            flop += numpy.prod([idx.extent for idx in node.multiindex + node.free_indices])
+        elif isinstance(node, (Terminal, Indexed, ListTensor, FlexiblyIndexed, LogicalOr, LogicalNot, LogicalAnd, Conditional)):
```

Comment:
This line deserves an `M-q`. (Emacs command that automatically wraps around long lines, breaking at the right length.)